In [1]:
from google import genai
from google.genai import types

print("Google GenAI SDK imported successfully!")

Google GenAI SDK imported successfully!


In [2]:
from dotenv import load_dotenv
import os

load_dotenv(r"D:\BECAME_DEVELOPER\hiver-sde-ai-agent\.env")

api_key = os.getenv("GEMINI_API_KEY")

print("API key loaded:", bool(api_key))

API key loaded: True


In [3]:
from google import genai

client = genai.Client(api_key=api_key)

print("Gemini client created successfully!")

Gemini client created successfully!


Test Gemini with ONE simple request

In [5]:
response = client.models.generate_content(
    model="gemini-3.6-flash",
    contents="Reply with exactly: Gemini API test successful."
)

print(response.text)

Gemini API test successful.


In [6]:
interaction = client.interactions.create(
    model="gemini-3.6-flash",
    input="How does AI work?"
)

print(interaction.output_text)

At its core, **Artificial Intelligence (AI) works by recognizing patterns in vast amounts of data and using those patterns to make predictions or decisions.** 

AI does not "think" or "feel" like a human. Instead, it uses mathematics, statistics, and computer code to process information and solve problems.

Here is a step-by-step breakdown of how AI works:

---

### 1. The Core Engine: The 4-Step Process

Every AI system relies on a simple four-part cycle:

1. **Input (Data):** The AI is fed huge amounts of data. This can be text, images, numbers, audio, or video. Data is the "fuel" that powers AI.
2. **Processing (Pattern Recognition):** The AI analyzes the data looking for similarities, trends, and rules. 
3. **Output (Prediction/Decision):** Based on the patterns it found, the AI makes a prediction or generates an answer. For example, *"Based on 10,000 photos of cats, there is a 98% chance this new photo is a cat."*
4. **Feedback (Learning):** Humans (or automated systems) tell the 

In [8]:
import pandas as pd

prompts_path = r"D:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\response_generation_prompts_150.csv"

generation_prompts = pd.read_csv(prompts_path)

print("Rows:", len(generation_prompts))
print("Columns:")
print(generation_prompts.columns.tolist())

Rows: 150
Columns:
['first_customer_message', 'last_customer_message', 'first_support_message', 'last_support_message', 'conversation_text', 'customer_turns', 'support_turns', 'conversation_size', 'start_time', 'end_time', 'duration_minutes', 'support_account', 'size_bucket', 'intent_label', 'annotation_notes', 'annotator', 'benchmark_id', 'customer_query', 'historical_conversation', '_match_text', 'matched_conversation_id', 'retrieval_benchmark_rows', 'retrieval_benchmark_index', 'conversation_id', 'retrieval_1_conversation_id', 'retrieval_1_score', 'retrieval_1_conversation_text', 'retrieval_1_first_customer', 'retrieval_1_last_customer', 'retrieval_1_first_support', 'retrieval_1_last_support', 'retrieval_1_support_account', 'retrieval_1_conversation_size', 'retrieval_2_conversation_id', 'retrieval_2_score', 'retrieval_2_conversation_text', 'retrieval_2_first_customer', 'retrieval_2_last_customer', 'retrieval_2_first_support', 'retrieval_2_last_support', 'retrieval_2_support_account'

In [9]:
row = generation_prompts.iloc[0]

print(row.to_dict())

{'first_customer_message': '@AmazonHelp please see the attached photos of the state of the ruined parcel which was left outside in the storm https://t.co/0DIvZDSRIP', 'last_customer_message': '@AmazonHelp Thank you', 'first_support_message': "@347202 I'm sorry for the condition of the parcel. Please contact us directly to explore options: https://t.co/JzP7hlA23B ^JF", 'last_support_message': "@347202 I'm sorry for the condition of the parcel. Please contact us directly to explore options: https://t.co/JzP7hlA23B ^JF", 'conversation_text': "[CUSTOMER] @AmazonHelp please see the attached photos of the state of the ruined parcel which was left outside in the storm https://t.co/0DIvZDSRIP\n[SUPPORT] @347202 I'm sorry for the condition of the parcel. Please contact us directly to explore options: https://t.co/JzP7hlA23B ^JF\n[CUSTOMER] @AmazonHelp Thank you", 'customer_turns': 2, 'support_turns': 1, 'conversation_size': 3, 'start_time': '2017-10-21 19:03:22+00:00', 'end_time': '2017-10-21 1

In [11]:
print(generation_prompts.columns.tolist())

['first_customer_message', 'last_customer_message', 'first_support_message', 'last_support_message', 'conversation_text', 'customer_turns', 'support_turns', 'conversation_size', 'start_time', 'end_time', 'duration_minutes', 'support_account', 'size_bucket', 'intent_label', 'annotation_notes', 'annotator', 'benchmark_id', 'customer_query', 'historical_conversation', '_match_text', 'matched_conversation_id', 'retrieval_benchmark_rows', 'retrieval_benchmark_index', 'conversation_id', 'retrieval_1_conversation_id', 'retrieval_1_score', 'retrieval_1_conversation_text', 'retrieval_1_first_customer', 'retrieval_1_last_customer', 'retrieval_1_first_support', 'retrieval_1_last_support', 'retrieval_1_support_account', 'retrieval_1_conversation_size', 'retrieval_2_conversation_id', 'retrieval_2_score', 'retrieval_2_conversation_text', 'retrieval_2_first_customer', 'retrieval_2_last_customer', 'retrieval_2_first_support', 'retrieval_2_last_support', 'retrieval_2_support_account', 'retrieval_2_conv

In [12]:
print(generation_prompts.head(1).T)

                                                                               0
first_customer_message         @AmazonHelp please see the attached photos of ...
last_customer_message                                      @AmazonHelp Thank you
first_support_message          @347202 I'm sorry for the condition of the par...
last_support_message           @347202 I'm sorry for the condition of the par...
conversation_text              [CUSTOMER] @AmazonHelp please see the attached...
customer_turns                                                                 2
support_turns                                                                  1
conversation_size                                                              3
start_time                                             2017-10-21 19:03:22+00:00
end_time                                               2017-10-21 19:31:10+00:00
duration_minutes                                                            27.8
support_account             

In [13]:
row = generation_prompts.iloc[0]

print("Benchmark ID:", row["benchmark_id"])
print("Intent:", row["intent_label"])
print("Policy:", row["policy_decision"])
print("Policy reason:", row["policy_reason_text"])
print("Policy rule:", row["policy_rule"])

print("\n========== GENERATION CONTEXT ==========\n")
print(row["generation_context"])

print("\n========== GENERATION PROMPT ==========\n")
print(row["generation_prompt"])

Benchmark ID: AUDIT_001
Intent: PRODUCT_PROBLEM
Policy: AUTO_HANDLE
Policy reason: Intent is actionable and sufficient historical evidence is available.
Policy rule: ACTIONABLE_WITH_EVIDENCE

========== GENERATION CONTEXT ==========

[CUSTOMER] @AmazonHelp please see the attached photos of the state of the ruined parcel which was left outside in the storm https://t.co/0DIvZDSRIP
[SUPPORT] @347202 I'm sorry for the condition of the parcel. Please contact us directly to explore options: https://t.co/JzP7hlA23B ^JF
[CUSTOMER] @AmazonHelp Thank you

========== GENERATION PROMPT ==========

You are an AI customer-support response assistant.

Draft a concise customer-facing support reply based on
the customer's full conversation context and the historical
support evidence provided below.

Use the full conversation to understand the customer's
current support state.

If the latest customer message is only an acknowledgement
such as "thank you", do not invent a new support request.

If the iss

In [14]:
row = generation_prompts.iloc[0]

prompt = row["generation_prompt"]

interaction = client.interactions.create(
    model="gemini-3.6-flash",
    input=prompt,
    store=False
)

gemini_reply = interaction.output_text

print(gemini_reply)

You're very welcome! Please don't hesitate to reach back out if you need anything else.


#### Generate all 150 responses

In [ ]:
import os
import time
import pandas as pd

output_path = r"D:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\response_generation_results_150.csv"
checkpoint_path = r"D:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\response_generation_checkpoint.csv"

results = []

for i, row in generation_prompts.iterrows():

    benchmark_id = row["benchmark_id"]

    print(f"[{i+1}/150] Generating {benchmark_id}...")

    try:
        interaction = client.interactions.create(
            model="gemini-3.6-flash",
            input=row["generation_prompt"],
            store=False
        )

        reply = interaction.output_text

        status = "SUCCESS"

    except Exception as e:
        print(f"  ERROR: {type(e).__name__}: {e}")
        reply = ""
        status = f"ERROR: {type(e).__name__}: {e}"

    results.append({
        "benchmark_id": benchmark_id,
        "conversation_id": row["conversation_id"],
        "intent_label": row["intent_label"],
        "policy_decision": row["policy_decision"],
        "policy_reason_text": row["policy_reason_text"],
        "policy_rule": row["policy_rule"],
        "gemini_reply": reply,
        "generation_status": status
    })

    # Checkpoint every 10 responses
    if (i + 1) % 10 == 0:
        checkpoint_df = pd.DataFrame(results)
        checkpoint_df.to_csv(
            checkpoint_path,
            index=False,
            encoding="utf-8"
        )

        print(f"  ✓ Checkpoint saved: {i+1}/150")

    time.sleep(0.5)

# Final save
results_df = pd.DataFrame(results)

results_df.to_csv(
    output_path,
    index=False,
    encoding="utf-8"
)

print("\n==============================")
print("GENERATION COMPLETE")
print("==============================")
print("Rows:", len(results_df))
print("Successful:", (results_df["generation_status"] == "SUCCESS").sum())
print("Errors:", (results_df["generation_status"] != "SUCCESS").sum())
print("Saved:", output_path)

In [17]:
from dotenv import load_dotenv
import os

load_dotenv(r"D:\BECAME_DEVELOPER\hiver-sde-ai-agent\.env")

groq_api_key = os.getenv("GROQ_API_KEY")

print("Groq API key loaded:", bool(groq_api_key))

Groq API key loaded: True


In [18]:
from groq import Groq

groq_client = Groq(api_key=groq_api_key)

print("Groq client created successfully!")

Groq client created successfully!


In [19]:
row = generation_prompts.iloc[3]

print("Benchmark:", row["benchmark_id"])
print("Intent:", row["intent_label"])
print("Policy:", row["policy_decision"])

Benchmark: AUDIT_004
Intent: DELIVERY_DELAY
Policy: AUTO_HANDLE


In [21]:
groq_response = groq_client.chat.completions.create(
    model="openai/gpt-oss-20b",
    messages=[
        {
            "role": "user",
            "content": row["generation_prompt"]
        }
    ],
    temperature=0.2,
    max_completion_tokens=300
)

groq_reply = groq_response.choices[0].message.content

print(groq_reply)

In [22]:
print("Finish reason:", groq_response.choices[0].finish_reason)
print("Message object:")
print(groq_response.choices[0].message)

print("\nContent:")
print(repr(groq_response.choices[0].message.content))

print("\nReasoning:")
print(repr(groq_response.choices[0].message.reasoning))

Finish reason: length
Message object:
ChatCompletionMessage(content='', role='assistant', annotations=None, executed_tools=None, function_call=None, reasoning='We need to draft a concise reply to the customer. The conversation: Customer says: "I had an issue with delivery. Ordered on 3rd. Product is still not delivered. No help from Amazon side. Just asking me to wait @115850". Support responded with email check. Customer says: "Yes, I got many emails regarding this issue and finally they want me to wait till tomorrow. Hope this will be resolved tomorrow." Support: "Thanks for confirming. Please be assured we shall resolve this at the earliest."\n\nNow the intent is DELIVERY_DELAY. Escalation decision: AUTO_HANDLE. Escalation reason: Intent actionable and sufficient historical evidence available. So we need to provide a response that addresses the delay, acknowledges the wait, and maybe provide next steps. According to rules, we cannot claim we performed action. We can say we will look

In [27]:
groq_response = groq_client.chat.completions.create(
    model="openai/gpt-oss-20b",
    messages=[
        {
            "role": "user",
            "content": "Reply with exactly: Groq test successful."
        }
    ],
    temperature=0.2,
    max_completion_tokens=1024,
    reasoning_effort="low",
    stream=False
)

print("Finish:", groq_response.choices[0].finish_reason)
print("Reply:", repr(groq_response.choices[0].message.content))
print("Usage:", groq_response.usage)

Finish: stop
Reply: 'Groq test successful.'
Usage: CompletionUsage(completion_tokens=21, prompt_tokens=80, total_tokens=101, completion_time=0.033500262, completion_tokens_details=CompletionTokensDetails(reasoning_tokens=7), prompt_time=0.016946638, prompt_tokens_details=None, queue_time=0.350184221, total_time=0.0504469)


In [28]:
# AUDIT_004
row = generation_prompts.iloc[3]

groq_response = groq_client.chat.completions.create(
    model="openai/gpt-oss-20b",
    messages=[
        {
            "role": "user",
            "content": row["generation_prompt"]
        }
    ],
    temperature=0.2,
    max_completion_tokens=1024,
    reasoning_effort="low",
    stream=False
)

groq_reply = groq_response.choices[0].message.content

print("Benchmark:", row["benchmark_id"])
print("Finish:", groq_response.choices[0].finish_reason)
print("\nGenerated reply:\n")
print(groq_reply)

Benchmark: AUDIT_004
Finish: stop

Generated reply:

I’m sorry you’re still waiting for your order.  
Please share your order number and the email address you used to place the order so we can look into the delay and get you an update as soon as possible. Thank you for your patience.


Provider: Groq
Model: openai/gpt-oss-20b
Reasoning: low
Temperature: 0.2
Max completion tokens: 1024

In [29]:
import os
import time
import pandas as pd

output_path = r"D:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\groq_response_generation_150.csv"
checkpoint_path = r"D:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\groq_response_generation_checkpoint.csv"

results = []

# Start from AUDIT_004
generation_subset = generation_prompts.iloc[3:].copy()

for i, (_, row) in enumerate(generation_subset.iterrows(), start=4):

    benchmark_id = row["benchmark_id"]

    print(f"[{i}/150] Generating {benchmark_id}...")

    try:
        response = groq_client.chat.completions.create(
            model="openai/gpt-oss-20b",
            messages=[
                {
                    "role": "user",
                    "content": row["generation_prompt"]
                }
            ],
            temperature=0.2,
            max_completion_tokens=1024,
            reasoning_effort="low",
            stream=False
        )

        reply = response.choices[0].message.content
        finish_reason = response.choices[0].finish_reason

        if not reply or not reply.strip():
            status = "EMPTY_RESPONSE"
        else:
            status = "SUCCESS"

        reasoning_tokens = getattr(
            response.usage.completion_tokens_details,
            "reasoning_tokens",
            None
        )

    except Exception as e:
        print(f"  ERROR: {type(e).__name__}: {e}")

        reply = ""
        finish_reason = ""
        reasoning_tokens = None
        status = f"ERROR: {type(e).__name__}: {e}"

    results.append({
        "benchmark_id": benchmark_id,
        "conversation_id": row["conversation_id"],
        "intent_label": row["intent_label"],
        "policy_decision": row["policy_decision"],
        "policy_reason_text": row["policy_reason_text"],
        "policy_rule": row["policy_rule"],
        "gemini_reply": "",
        "groq_reply": reply,
        "provider": "groq",
        "model": "openai/gpt-oss-20b",
        "temperature": 0.2,
        "reasoning_effort": "low",
        "max_completion_tokens": 1024,
        "finish_reason": finish_reason,
        "reasoning_tokens": reasoning_tokens,
        "generation_status": status
    })

    # Save every 10 examples
    if i % 10 == 0 or i == 150:
        checkpoint_df = pd.DataFrame(results)

        checkpoint_df.to_csv(
            checkpoint_path,
            index=False,
            encoding="utf-8"
        )

        print(f"  ✓ Checkpoint saved: {i}/150")

    time.sleep(0.5)

results_df = pd.DataFrame(results)

results_df.to_csv(
    output_path,
    index=False,
    encoding="utf-8"
)

print("\n==============================")
print("GROQ GENERATION COMPLETE")
print("==============================")
print("Rows:", len(results_df))
print("Successful:", (results_df["generation_status"] == "SUCCESS").sum())
print("Empty:", (results_df["generation_status"] == "EMPTY_RESPONSE").sum())
print("Errors:", results_df["generation_status"].str.startswith("ERROR").sum())
print("Saved:", output_path)

[4/150] Generating AUDIT_004...
[5/150] Generating AUDIT_005...
[6/150] Generating AUDIT_006...
[7/150] Generating AUDIT_007...
[8/150] Generating AUDIT_008...
[9/150] Generating AUDIT_009...
[10/150] Generating AUDIT_010...
  ✓ Checkpoint saved: 10/150
[11/150] Generating AUDIT_011...
[12/150] Generating AUDIT_012...
[13/150] Generating AUDIT_013...
[14/150] Generating AUDIT_014...
[15/150] Generating AUDIT_015...
[16/150] Generating AUDIT_016...
[17/150] Generating AUDIT_017...
[18/150] Generating AUDIT_018...
[19/150] Generating AUDIT_019...
[20/150] Generating AUDIT_020...
  ✓ Checkpoint saved: 20/150
[21/150] Generating AUDIT_021...
[22/150] Generating AUDIT_022...
[23/150] Generating AUDIT_023...
[24/150] Generating AUDIT_024...
[25/150] Generating AUDIT_025...
[26/150] Generating AUDIT_026...
[27/150] Generating AUDIT_027...
[28/150] Generating AUDIT_028...
[29/150] Generating AUDIT_029...
[30/150] Generating AUDIT_030...
  ✓ Checkpoint saved: 30/150
[31/150] Generating AUDIT_03

In [30]:
import pandas as pd

first_three_results = []

for i in range(3):
    row = generation_prompts.iloc[i]

    print(f"[{i+1}/3] Generating {row['benchmark_id']}...")

    try:
        response = groq_client.chat.completions.create(
            model="openai/gpt-oss-20b",
            messages=[
                {
                    "role": "user",
                    "content": row["generation_prompt"]
                }
            ],
            temperature=0.2,
            max_completion_tokens=1024,
            reasoning_effort="low",
            stream=False
        )

        reply = response.choices[0].message.content
        finish_reason = response.choices[0].finish_reason

        reasoning_tokens = getattr(
            response.usage.completion_tokens_details,
            "reasoning_tokens",
            None
        )

        status = "SUCCESS" if reply and reply.strip() else "EMPTY_RESPONSE"

    except Exception as e:
        print(f"  ERROR: {type(e).__name__}: {e}")

        reply = ""
        finish_reason = ""
        reasoning_tokens = None
        status = f"ERROR: {type(e).__name__}: {e}"

    first_three_results.append({
        "benchmark_id": row["benchmark_id"],
        "conversation_id": row["conversation_id"],
        "intent_label": row["intent_label"],
        "policy_decision": row["policy_decision"],
        "policy_reason_text": row["policy_reason_text"],
        "policy_rule": row["policy_rule"],
        "gemini_reply": "",
        "groq_reply": reply,
        "provider": "groq",
        "model": "openai/gpt-oss-20b",
        "temperature": 0.2,
        "reasoning_effort": "low",
        "max_completion_tokens": 1024,
        "finish_reason": finish_reason,
        "reasoning_tokens": reasoning_tokens,
        "generation_status": status
    })

first_three_df = pd.DataFrame(first_three_results)

print("\nGenerated:", len(first_three_df))
print("Successful:", (first_three_df["generation_status"] == "SUCCESS").sum())
print("Empty:", (first_three_df["generation_status"] == "EMPTY_RESPONSE").sum())
print(
    "Errors:",
    first_three_df["generation_status"].str.startswith("ERROR").sum()
)

[1/3] Generating AUDIT_001...
[2/3] Generating AUDIT_002...
[3/3] Generating AUDIT_003...

Generated: 3
Successful: 3
Empty: 0
Errors: 0


In [31]:
groq_existing_path = r"D:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\groq_response_generation_150.csv"

groq_existing_df = pd.read_csv(groq_existing_path)

print("Existing rows:", len(groq_existing_df))

Existing rows: 147


In [32]:
groq_full_df = pd.concat(
    [first_three_df, groq_existing_df],
    ignore_index=True
)

groq_full_df = groq_full_df.sort_values(
    "benchmark_id"
).reset_index(drop=True)

print("Total rows:", len(groq_full_df))
print("Unique benchmark IDs:", groq_full_df["benchmark_id"].nunique())

print("\nStatus:")
print(groq_full_df["generation_status"].value_counts())

print("\nProviders:")
print(groq_full_df["provider"].value_counts())

Total rows: 150
Unique benchmark IDs: 150

Status:
generation_status
SUCCESS    150
Name: count, dtype: int64

Providers:
provider
groq    150
Name: count, dtype: int64


In [33]:
final_path = r"D:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\groq_response_generation_150.csv"

groq_full_df.to_csv(
    final_path,
    index=False,
    encoding="utf-8"
)

print("Final Groq benchmark saved:")
print(final_path)

Final Groq benchmark saved:
D:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\groq_response_generation_150.csv


In [34]:
assert len(groq_full_df) == 150
assert groq_full_df["benchmark_id"].nunique() == 150
assert (groq_full_df["provider"] == "groq").all()
assert (groq_full_df["generation_status"] == "SUCCESS").all()
assert groq_full_df["groq_reply"].notna().all()
assert groq_full_df["groq_reply"].str.strip().ne("").all()

print("✅ 150/150 Groq response-generation benchmark validated.")

✅ 150/150 Groq response-generation benchmark validated.


In [40]:
import pandas as pd
from pathlib import Path

benchmark_path = Path(
    r"d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\response_generation_prompts_150.csv"
)

benchmark_df = pd.read_csv(benchmark_path)

print("Benchmark shape:", benchmark_df.shape)
print("\nRequired columns:")
required_cols = [
    "benchmark_id",
    "conversation_text",
    "intent_label",
    "policy_decision",
    "policy_reason_text",
    "retrieval_1_conversation_text",
    "retrieval_2_conversation_text",
    "retrieval_3_conversation_text",
]

for col in required_cols:
    print(f"{col}: {col in benchmark_df.columns}")

Benchmark shape: (150, 57)

Required columns:
benchmark_id: True
conversation_text: True
intent_label: True
policy_decision: True
policy_reason_text: True
retrieval_1_conversation_text: True
retrieval_2_conversation_text: True
retrieval_3_conversation_text: True


In [41]:
# Merge the original benchmark context with the existing Groq responses

judge_df = benchmark_df.merge(
    groq_full_df[
        [
            "benchmark_id",
            "groq_reply",
            "provider",
            "model",
            "generation_status"
        ]
    ],
    on="benchmark_id",
    how="inner",
    validate="one_to_one"
)

print("Judge dataframe shape:", judge_df.shape)

print("\nUnique benchmark IDs:", judge_df["benchmark_id"].nunique())
print("Missing Groq replies:", judge_df["groq_reply"].isna().sum())
print("Generation failures:", (judge_df["generation_status"] != "SUCCESS").sum())

print("\nRequired judge columns:")
judge_required = [
    "conversation_text",
    "intent_label",
    "policy_decision",
    "policy_reason_text",
    "retrieval_1_conversation_text",
    "retrieval_2_conversation_text",
    "retrieval_3_conversation_text",
    "groq_reply"
]

for col in judge_required:
    print(f"{col}: {col in judge_df.columns}")

Judge dataframe shape: (150, 61)

Unique benchmark IDs: 150
Missing Groq replies: 0
Generation failures: 0

Required judge columns:
conversation_text: True
intent_label: True
policy_decision: True
policy_reason_text: True
retrieval_1_conversation_text: True
retrieval_2_conversation_text: True
retrieval_3_conversation_text: True
groq_reply: True


#### Create the judge prompt

In [42]:
def build_judge_prompt(row):
    return f"""
You are evaluating an AI customer-support response.

Evaluate the generated response using ONLY the customer conversation,
the historical support evidence, the stated intent, and the stated
escalation decision provided below.

Do not assume policies, facts, or actions that are not supported by
the evidence.

CUSTOMER CONVERSATION:
{row["conversation_text"]}

INTENT:
{row["intent_label"]}

ESCALATION DECISION:
{row["policy_decision"]}

ESCALATION REASON:
{row["policy_reason_text"]}

HISTORICAL SUPPORT EVIDENCE:

EPISODE 1:
{row["retrieval_1_conversation_text"]}

EPISODE 2:
{row["retrieval_2_conversation_text"]}

EPISODE 3:
{row["retrieval_3_conversation_text"]}

GENERATED RESPONSE:
{row["groq_reply"]}

Score each dimension from 0 to 2.

CORRECTNESS:
0 = incorrect, misleading, or contradicts the evidence
1 = partially correct or materially incomplete
2 = correct and consistent with the evidence

RELEVANCE:
0 = does not address the customer's actual issue
1 = partially addresses the issue
2 = directly addresses the customer's actual issue

GROUNDING:
0 = contains unsupported or fabricated claims/actions
1 = mostly grounded but contains minor unsupported detail
2 = fully grounded in the supplied conversation/evidence

HELPFULNESS:
0 = provides no useful assistance
1 = somewhat useful
2 = provides an appropriate useful next step or response

TONE:
0 = inappropriate, unprofessional, or insensitive
1 = acceptable
2 = professional, concise, and appropriately empathetic

COMPLETENESS:
0 = misses the central support situation
1 = addresses it partially
2 = adequately handles the support situation given the available evidence

FAILURE FLAGS:
Select any applicable flags:
- FABRICATED_DETAIL
- UNSUPPORTED_ACTION
- WRONG_INTENT
- MISSED_ESCALATION
- UNNECESSARY_ESCALATION
- IRRELEVANT_RESPONSE
- INCOMPLETE_RESPONSE
- NONE

Return ONLY valid JSON in exactly this structure:

{{
  "correctness": 0,
  "relevance": 0,
  "grounding": 0,
  "helpfulness": 0,
  "tone": 0,
  "completeness": 0,
  "failure_flags": [],
  "brief_reason": ""
}}
"""

In [37]:
print("Shape:", groq_full_df.shape)
print("\nColumns:")
for col in groq_full_df.columns:
    print(col)

Shape: (150, 16)

Columns:
benchmark_id
conversation_id
intent_label
policy_decision
policy_reason_text
policy_rule
gemini_reply
groq_reply
provider
model
temperature
reasoning_effort
max_completion_tokens
finish_reason
reasoning_tokens
generation_status


In [38]:
print(groq_full_df.head(1).T)

                                                                       0
benchmark_id                                                   AUDIT_001
conversation_id                                                   957549
intent_label                                             PRODUCT_PROBLEM
policy_decision                                              AUTO_HANDLE
policy_reason_text     Intent is actionable and sufficient historical...
policy_rule                                     ACTIONABLE_WITH_EVIDENCE
gemini_reply                                                            
groq_reply             Hi there,\n\nI’m sorry you’ve had to deal with...
provider                                                            groq
model                                                 openai/gpt-oss-20b
temperature                                                          0.2
reasoning_effort                                                     low
max_completion_tokens                              

In [43]:
judge_test_results = []

for i in range(5):
    row = judge_df.iloc[i]

    judge_prompt = build_judge_prompt(row)

    response = groq_client.chat.completions.create(
        model="openai/gpt-oss-20b",
        messages=[
            {
                "role": "user",
                "content": judge_prompt
            }
        ],
        temperature=0.0,
        max_completion_tokens=1024,
        reasoning_effort="low",
        stream=False
    )

    judge_output = response.choices[0].message.content

    result = {
        "benchmark_id": row["benchmark_id"],
        "finish_reason": response.choices[0].finish_reason,
        "reasoning_tokens": getattr(
            response.usage,
            "completion_tokens_details",
            None
        ),
        "judge_output": judge_output
    }

    judge_test_results.append(result)

    print("=" * 80)
    print("Benchmark:", row["benchmark_id"])
    print("Finish reason:", result["finish_reason"])
    print("Judge output:")
    print(judge_output)

Benchmark: AUDIT_001
Finish reason: stop
Judge output:
{
  "correctness": 2,
  "relevance": 2,
  "grounding": 2,
  "helpfulness": 2,
  "tone": 2,
  "completeness": 2,
  "failure_flags": ["NONE"],
  "brief_reason": "The response accurately reflects the customer’s issue, aligns with the historical evidence, offers a clear next step, and maintains a professional, empathetic tone."
}
Benchmark: AUDIT_002
Finish reason: stop
Judge output:
{
  "correctness": 0,
  "relevance": 0,
  "grounding": 0,
  "helpfulness": 0,
  "tone": 1,
  "completeness": 0,
  "failure_flags": [
    "MISSED_ESCALATION",
    "IRRELEVANT_RESPONSE",
    "INCOMPLETE_RESPONSE"
  ],
  "brief_reason": "The response does not address the customer’s comment about packaging, contains no evidence-based content, and fails to acknowledge the required escalation. It is therefore incorrect, irrelevant, ungrounded, unhelpful, and incomplete."
}
Benchmark: AUDIT_003
Finish reason: stop
Judge output:
{
  "correctness": 2,
  "relevance"

In [44]:
import json
import time
import pandas as pd

judge_results = []

for i, row in judge_df.reset_index(drop=True).iterrows():

    judge_prompt = build_judge_prompt(row)

    try:
        response = groq_client.chat.completions.create(
            model="openai/gpt-oss-20b",
            messages=[
                {
                    "role": "user",
                    "content": judge_prompt
                }
            ],
            temperature=0.0,
            max_completion_tokens=1024,
            reasoning_effort="low",
            stream=False
        )

        judge_output = response.choices[0].message.content

        judge_results.append({
            "benchmark_id": row["benchmark_id"],
            "judge_output": judge_output,
            "finish_reason": response.choices[0].finish_reason,
            "judge_status": "SUCCESS"
        })

    except Exception as e:
        judge_results.append({
            "benchmark_id": row["benchmark_id"],
            "judge_output": "",
            "finish_reason": "",
            "judge_status": f"ERROR: {str(e)}"
        })

    if (i + 1) % 10 == 0:
        print(f"Completed {i + 1}/150")

    time.sleep(0.15)

judge_raw_df = pd.DataFrame(judge_results)

print("\n" + "=" * 60)
print("Judge run complete")
print("Rows:", len(judge_raw_df))
print("Success:", (judge_raw_df["judge_status"] == "SUCCESS").sum())
print("Errors:", (judge_raw_df["judge_status"] != "SUCCESS").sum())
print("Non-stop finishes:")
print((judge_raw_df["finish_reason"] != "stop").sum())

Completed 10/150
Completed 20/150
Completed 30/150
Completed 40/150
Completed 50/150
Completed 60/150
Completed 70/150
Completed 80/150
Completed 90/150
Completed 100/150
Completed 110/150
Completed 120/150
Completed 130/150
Completed 140/150
Completed 150/150

Judge run complete
Rows: 150
Success: 28
Errors: 122
Non-stop finishes:
122


In [45]:
# Inspect the actual errors from the failed judge calls

failed_judges = judge_raw_df[
    judge_raw_df["judge_status"] != "SUCCESS"
].copy()

print("Failed rows:", len(failed_judges))
print("\nUnique error messages:\n")

for error, count in failed_judges["judge_status"].value_counts().items():
    print(f"{count} × {error}")

Failed rows: 122

Unique error messages:

1 × ERROR: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-20b` in organization `org_01kwpae1wpf53rprz1pnt1dv5y` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 198975, Requested 2098. Please try again in 7m43.536s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
1 × ERROR: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-20b` in organization `org_01kwpae1wpf53rprz1pnt1dv5y` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 199975, Requested 1881. Please try again in 13m21.792s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
1 × ERROR: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-20b` in organization `org_01

In [46]:
# Preserve the successful judge results and identify exactly what remains

successful_judges = judge_raw_df[
    judge_raw_df["judge_status"] == "SUCCESS"
].copy()

successful_ids = set(successful_judges["benchmark_id"])

remaining_judge_df = judge_df[
    ~judge_df["benchmark_id"].isin(successful_ids)
].copy()

print("Successful judgments:", len(successful_judges))
print("Remaining judgments:", len(remaining_judge_df))
print("Total benchmark rows:", len(judge_df))

print("\nSuccessful IDs:")
print(sorted(successful_ids))

print("\nFirst remaining IDs:")
print(remaining_judge_df["benchmark_id"].head(20).tolist())

Successful judgments: 28
Remaining judgments: 122
Total benchmark rows: 150

Successful IDs:
['AUDIT_001', 'AUDIT_002', 'AUDIT_003', 'AUDIT_004', 'AUDIT_005', 'AUDIT_006', 'AUDIT_007', 'AUDIT_008', 'AUDIT_009', 'AUDIT_010', 'AUDIT_011', 'AUDIT_012', 'AUDIT_013', 'AUDIT_014', 'AUDIT_015', 'AUDIT_016', 'AUDIT_017', 'AUDIT_018', 'AUDIT_019', 'AUDIT_020', 'AUDIT_021', 'AUDIT_022', 'AUDIT_023', 'AUDIT_024', 'AUDIT_025', 'AUDIT_026', 'AUDIT_027', 'AUDIT_072']

First remaining IDs:
['AUDIT_028', 'AUDIT_029', 'AUDIT_030', 'AUDIT_031', 'AUDIT_032', 'AUDIT_033', 'AUDIT_034', 'AUDIT_035', 'AUDIT_036', 'AUDIT_037', 'AUDIT_038', 'AUDIT_039', 'AUDIT_040', 'AUDIT_041', 'AUDIT_042', 'AUDIT_043', 'AUDIT_044', 'AUDIT_045', 'AUDIT_046', 'AUDIT_047']


In [47]:
from pathlib import Path

checkpoint_path = Path(
    r"d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\llm_judge_checkpoint_28.csv"
)

successful_judges.to_csv(
    checkpoint_path,
    index=False,
    encoding="utf-8-sig"
)

print("Saved:", checkpoint_path)
print("Rows saved:", len(successful_judges))

Saved: d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\llm_judge_checkpoint_28.csv
Rows saved: 28


In [48]:
import json
import re
import pandas as pd
from pathlib import Path

def parse_judge_json(raw_output):
    """
    Parse an LLM judge response into structured evaluation fields.
    Returns a dictionary with None values if parsing fails.
    """

    result = {
        "correctness": None,
        "relevance": None,
        "grounding": None,
        "helpfulness": None,
        "tone": None,
        "completeness": None,
        "failure_flags": None,
        "brief_reason": None,
        "parse_status": "FAILED"
    }

    if not isinstance(raw_output, str) or not raw_output.strip():
        return result

    text = raw_output.strip()

    # Remove Markdown code fences if the model ever returns them.
    text = re.sub(r"^```(?:json)?\s*", "", text, flags=re.IGNORECASE)
    text = re.sub(r"\s*```$", "", text)

    # First attempt: parse the entire response.
    try:
        data = json.loads(text)
    except json.JSONDecodeError:

        # Second attempt: extract the first JSON object.
        match = re.search(r"\{.*\}", text, flags=re.DOTALL)

        if not match:
            return result

        try:
            data = json.loads(match.group(0))
        except json.JSONDecodeError:
            return result

    score_fields = [
        "correctness",
        "relevance",
        "grounding",
        "helpfulness",
        "tone",
        "completeness"
    ]

    valid = True

    for field in score_fields:
        value = data.get(field)

        if isinstance(value, bool):
            valid = False
            break

        try:
            value = int(value)
        except (TypeError, ValueError):
            valid = False
            break

        if value not in [0, 1, 2]:
            valid = False
            break

        result[field] = value

    if not valid:
        return result

    flags = data.get("failure_flags", [])

    if isinstance(flags, str):
        flags = [flags]

    if not isinstance(flags, list):
        return result

    result["failure_flags"] = flags
    result["brief_reason"] = str(data.get("brief_reason", "")).strip()
    result["parse_status"] = "SUCCESS"

    return result

In [49]:
parsed_test = judge_raw_df.copy()

parsed_columns = parsed_test["judge_output"].apply(parse_judge_json)

parsed_test = pd.concat(
    [
        parsed_test.drop(columns=["judge_output"]),
        pd.DataFrame(parsed_columns.tolist(), index=parsed_test.index)
    ],
    axis=1
)

print("Total judge results:", len(parsed_test))
print(
    "Successfully parsed:",
    (parsed_test["parse_status"] == "SUCCESS").sum()
)
print(
    "Parse failures:",
    (parsed_test["parse_status"] != "SUCCESS").sum()
)

print("\nParsed score preview:")
print(
    parsed_test[
        [
            "benchmark_id",
            "correctness",
            "relevance",
            "grounding",
            "helpfulness",
            "tone",
            "completeness",
            "failure_flags",
            "parse_status"
        ]
    ].head(10).to_string(index=False)
)

Total judge results: 150
Successfully parsed: 28
Parse failures: 122

Parsed score preview:
benchmark_id  correctness  relevance  grounding  helpfulness  tone  completeness                                                 failure_flags parse_status
   AUDIT_001          2.0        2.0        2.0          2.0   2.0           2.0                                                        [NONE]      SUCCESS
   AUDIT_002          0.0        0.0        0.0          0.0   1.0           0.0 [MISSED_ESCALATION, IRRELEVANT_RESPONSE, INCOMPLETE_RESPONSE]      SUCCESS
   AUDIT_003          2.0        2.0        2.0          2.0   2.0           2.0                                                            []      SUCCESS
   AUDIT_004          0.0        2.0        0.0          1.0   2.0           1.0                       [FABRICATED_DETAIL, UNSUPPORTED_ACTION]      SUCCESS
   AUDIT_005          2.0        2.0        2.0          2.0   2.0           2.0                                                

In [50]:
# Build the evaluation-ready dataframe.
# No API calls are made.

judge_scores = parsed_test[
    [
        "benchmark_id",
        "correctness",
        "relevance",
        "grounding",
        "helpfulness",
        "tone",
        "completeness",
        "failure_flags",
        "brief_reason",
        "parse_status"
    ]
].copy()

evaluation_df = judge_df.merge(
    judge_scores,
    on="benchmark_id",
    how="left",
    validate="one_to_one"
)

evaluation_df["judge_completed"] = (
    evaluation_df["parse_status"] == "SUCCESS"
)

print("Evaluation dataframe:", evaluation_df.shape)
print()
print("Generated responses:",
      evaluation_df["groq_reply"].notna().sum())
print("Judge completed:",
      evaluation_df["judge_completed"].sum())
print("Judge pending:",
      (~evaluation_df["judge_completed"]).sum())
print()
print("Score coverage:")
for col in [
    "correctness",
    "relevance",
    "grounding",
    "helpfulness",
    "tone",
    "completeness"
]:
    print(
        f"{col}: "
        f"{evaluation_df[col].notna().sum()}/150"
    )

Evaluation dataframe: (150, 71)

Generated responses: 150
Judge completed: 28
Judge pending: 122

Score coverage:
correctness: 28/150
relevance: 28/150
grounding: 28/150
helpfulness: 28/150
tone: 28/150
completeness: 28/150


In [51]:
completed_eval = evaluation_df[
    evaluation_df["judge_completed"]
].copy()

score_cols = [
    "correctness",
    "relevance",
    "grounding",
    "helpfulness",
    "tone",
    "completeness"
]

print("=" * 70)
print("INTERIM LLM-AS-JUDGE DIAGNOSTIC")
print("IMPORTANT: Based on 28 judged examples only")
print("=" * 70)

print("\nMean scores (0–2):")
for col in score_cols:
    print(
        f"{col:15s}: "
        f"{completed_eval[col].mean():.3f}"
    )

print("\nMean normalized scores (% of maximum):")
for col in score_cols:
    print(
        f"{col:15s}: "
        f"{completed_eval[col].mean() / 2 * 100:.1f}%"
    )

print("\nOverall mean score:")
overall_mean = completed_eval[score_cols].mean().mean()
print(f"{overall_mean:.3f} / 2")
print(f"{overall_mean / 2 * 100:.1f}%")

print("\nScore distributions:")
for col in score_cols:
    print(f"\n{col.upper()}")
    print(
        completed_eval[col]
        .value_counts()
        .sort_index()
        .to_dict()
    )

INTERIM LLM-AS-JUDGE DIAGNOSTIC
IMPORTANT: Based on 28 judged examples only

Mean scores (0–2):
correctness    : 1.571
relevance      : 1.571
grounding      : 1.607
helpfulness    : 1.536
tone           : 1.857
completeness   : 1.536

Mean normalized scores (% of maximum):
correctness    : 78.6%
relevance      : 78.6%
grounding      : 80.4%
helpfulness    : 76.8%
tone           : 92.9%
completeness   : 76.8%

Overall mean score:
1.613 / 2
80.7%

Score distributions:

CORRECTNESS
{0.0: 5, 1.0: 2, 2.0: 21}

RELEVANCE
{0.0: 4, 1.0: 4, 2.0: 20}

GROUNDING
{0.0: 5, 1.0: 1, 2.0: 22}

HELPFULNESS
{0.0: 4, 1.0: 5, 2.0: 19}

TONE
{1.0: 4, 2.0: 24}

COMPLETENESS
{0.0: 4, 1.0: 5, 2.0: 19}


In [52]:
from collections import Counter

failure_counter = Counter()

for flags in completed_eval["failure_flags"]:
    if isinstance(flags, list):
        for flag in flags:
            failure_counter[flag] += 1

print("=" * 70)
print("INTERIM FAILURE-MODE COUNTS")
print("Based on 28 judged examples")
print("=" * 70)

for flag, count in failure_counter.most_common():
    print(f"{flag:25s}: {count}")

print("\nExamples with failures:")
failure_rows = completed_eval[
    completed_eval["failure_flags"].apply(
        lambda x: isinstance(x, list) and len(x) > 0 and x != ["NONE"]
    )
]

print(
    failure_rows[
        [
            "benchmark_id",
            "intent_label",
            "policy_decision",
            "failure_flags",
            "brief_reason"
        ]
    ].to_string(index=False)
)

INTERIM FAILURE-MODE COUNTS
Based on 28 judged examples
MISSED_ESCALATION        : 4
INCOMPLETE_RESPONSE      : 4
NONE                     : 3
FABRICATED_DETAIL        : 3
UNSUPPORTED_ACTION       : 3
IRRELEVANT_RESPONSE      : 2
UNNECESSARY_ESCALATION   : 1

Examples with failures:
benchmark_id                     intent_label policy_decision                                                                      failure_flags                                                                                                                                                                                                                                                                                                                                                                                             brief_reason
   AUDIT_002                OTHER_NON_SUPPORT        ESCALATE                      [MISSED_ESCALATION, IRRELEVANT_RESPONSE, INCOMPLETE_RESPONSE]                                  

In [53]:
from pathlib import Path

pending_judge_path = Path(
    r"d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\llm_judge_pending_122.csv"
)

remaining_judge_df.to_csv(
    pending_judge_path,
    index=False,
    encoding="utf-8-sig"
)

print("Pending judge file saved:")
print(pending_judge_path)

print("\nRows:", len(remaining_judge_df))
print(
    "Unique benchmark IDs:",
    remaining_judge_df["benchmark_id"].nunique()
)

print(
    "\nRange:",
    remaining_judge_df["benchmark_id"].iloc[0],
    "→",
    remaining_judge_df["benchmark_id"].iloc[-1]
)


Pending judge file saved:
d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\llm_judge_pending_122.csv

Rows: 122
Unique benchmark IDs: 122

Range: AUDIT_028 → AUDIT_150


In [54]:
import time
import pandas as pd
from pathlib import Path

resume_output_path = Path(
    r"d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\llm_judge_resume_results.csv"
)

def run_remaining_judges(
    pending_df,
    output_path=resume_output_path,
    delay_seconds=0.2
):
    """
    Resume-safe LLM-as-judge runner.

    Makes API calls only for pending rows.
    Saves every successful result immediately.
    Stops on rate-limit errors so progress is never wasted.
    """

    # Load previous resume results if they already exist.
    if output_path.exists():
        saved_results = pd.read_csv(output_path)
        completed_ids = set(saved_results["benchmark_id"].astype(str))
        print(f"Existing resume results: {len(saved_results)}")
    else:
        saved_results = pd.DataFrame()
        completed_ids = set()
        print("No previous resume results found.")

    pending = pending_df[
        ~pending_df["benchmark_id"].astype(str).isin(completed_ids)
    ].copy()

    print(f"Remaining API calls needed: {len(pending)}")

    for idx, (_, row) in enumerate(pending.iterrows(), start=1):

        benchmark_id = str(row["benchmark_id"])

        try:
            judge_prompt = build_judge_prompt(row)

            response = groq_client.chat.completions.create(
                model="openai/gpt-oss-20b",
                messages=[
                    {
                        "role": "user",
                        "content": judge_prompt
                    }
                ],
                temperature=0.0,
                max_completion_tokens=1024,
                reasoning_effort="low",
                stream=False
            )

            judge_output = response.choices[0].message.content

            new_result = pd.DataFrame([{
                "benchmark_id": benchmark_id,
                "judge_output": judge_output,
                "finish_reason": response.choices[0].finish_reason,
                "judge_status": "SUCCESS"
            }])

            saved_results = pd.concat(
                [saved_results, new_result],
                ignore_index=True
            )

            # Save immediately after every successful judgment.
            saved_results.to_csv(
                output_path,
                index=False,
                encoding="utf-8-sig"
            )

            print(
                f"[{idx}/{len(pending)}] "
                f"{benchmark_id} ✓"
            )

            time.sleep(delay_seconds)

        except Exception as e:

            error_text = str(e)

            print(
                f"\nSTOPPED at {benchmark_id}"
            )
            print(error_text)

            # Do not continue after a rate-limit error.
            if "429" in error_text or "rate_limit" in error_text.lower():
                print("\nRate limit detected.")
                print("Successful results have already been saved.")
                break

            # For another unexpected error, also stop rather than
            # silently skipping a benchmark.
            raise

    print("\n" + "=" * 60)
    print("RESUME RUN STATUS")
    print("=" * 60)
    print("Saved successful judgments:", len(saved_results))
    print("Output:", output_path)

    return saved_results

In [55]:
from pathlib import Path
import pandas as pd

checkpoint_path = Path(
    r"d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\llm_judge_checkpoint_28.csv"
)

pending_path = Path(
    r"d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\llm_judge_pending_122.csv"
)

checkpoint_df = pd.read_csv(checkpoint_path)
pending_df = pd.read_csv(pending_path)

checkpoint_ids = set(checkpoint_df["benchmark_id"].astype(str))
pending_ids = set(pending_df["benchmark_id"].astype(str))
all_ids = set(benchmark_df["benchmark_id"].astype(str))

overlap = sorted(checkpoint_ids & pending_ids)
missing = sorted(all_ids - checkpoint_ids - pending_ids)
extra = sorted((checkpoint_ids | pending_ids) - all_ids)

print("Checkpoint rows:", len(checkpoint_df))
print("Checkpoint unique IDs:", len(checkpoint_ids))

print("Pending rows:", len(pending_df))
print("Pending unique IDs:", len(pending_ids))

print("\nOverlap:", overlap)
print("Missing benchmark IDs:", missing)
print("Extra IDs:", extra)

print("\nUnion unique IDs:", len(checkpoint_ids | pending_ids))
print("Expected benchmark IDs:", len(all_ids))

Checkpoint rows: 28
Checkpoint unique IDs: 28
Pending rows: 122
Pending unique IDs: 122

Overlap: []
Missing benchmark IDs: []
Extra IDs: []

Union unique IDs: 150
Expected benchmark IDs: 150


In [56]:
resume_results_df = run_remaining_judges(
    pending_df=pending_df,
    output_path=resume_output_path,
    delay_seconds=0.2
)

No previous resume results found.
Remaining API calls needed: 122
[1/122] AUDIT_028 ✓
[2/122] AUDIT_029 ✓
[3/122] AUDIT_030 ✓
[4/122] AUDIT_031 ✓
[5/122] AUDIT_032 ✓
[6/122] AUDIT_033 ✓
[7/122] AUDIT_034 ✓
[8/122] AUDIT_035 ✓
[9/122] AUDIT_036 ✓
[10/122] AUDIT_037 ✓
[11/122] AUDIT_038 ✓
[12/122] AUDIT_039 ✓
[13/122] AUDIT_040 ✓
[14/122] AUDIT_041 ✓
[15/122] AUDIT_042 ✓
[16/122] AUDIT_043 ✓
[17/122] AUDIT_044 ✓
[18/122] AUDIT_045 ✓
[19/122] AUDIT_046 ✓
[20/122] AUDIT_047 ✓
[21/122] AUDIT_048 ✓
[22/122] AUDIT_049 ✓
[23/122] AUDIT_050 ✓
[24/122] AUDIT_051 ✓
[25/122] AUDIT_052 ✓
[26/122] AUDIT_053 ✓
[27/122] AUDIT_054 ✓
[28/122] AUDIT_055 ✓
[29/122] AUDIT_056 ✓
[30/122] AUDIT_057 ✓
[31/122] AUDIT_058 ✓
[32/122] AUDIT_059 ✓
[33/122] AUDIT_060 ✓
[34/122] AUDIT_061 ✓
[35/122] AUDIT_062 ✓
[36/122] AUDIT_063 ✓
[37/122] AUDIT_064 ✓
[38/122] AUDIT_065 ✓
[39/122] AUDIT_066 ✓
[40/122] AUDIT_067 ✓
[41/122] AUDIT_068 ✓
[42/122] AUDIT_069 ✓
[43/122] AUDIT_070 ✓

STOPPED at AUDIT_071
Error code: 429 - 

In [57]:
resume_results_df = run_remaining_judges(
    pending_df=pending_df,
    output_path=resume_output_path,
    delay_seconds=0.2
)

Existing resume results: 43
Remaining API calls needed: 79
[1/79] AUDIT_071 ✓
[2/79] AUDIT_073 ✓
[3/79] AUDIT_074 ✓
[4/79] AUDIT_075 ✓
[5/79] AUDIT_076 ✓
[6/79] AUDIT_077 ✓
[7/79] AUDIT_078 ✓
[8/79] AUDIT_079 ✓
[9/79] AUDIT_080 ✓
[10/79] AUDIT_081 ✓
[11/79] AUDIT_082 ✓
[12/79] AUDIT_083 ✓
[13/79] AUDIT_084 ✓
[14/79] AUDIT_085 ✓
[15/79] AUDIT_086 ✓
[16/79] AUDIT_087 ✓
[17/79] AUDIT_088 ✓
[18/79] AUDIT_089 ✓
[19/79] AUDIT_090 ✓
[20/79] AUDIT_091 ✓
[21/79] AUDIT_092 ✓
[22/79] AUDIT_093 ✓
[23/79] AUDIT_094 ✓
[24/79] AUDIT_095 ✓
[25/79] AUDIT_096 ✓
[26/79] AUDIT_097 ✓
[27/79] AUDIT_098 ✓
[28/79] AUDIT_099 ✓
[29/79] AUDIT_100 ✓
[30/79] AUDIT_101 ✓
[31/79] AUDIT_102 ✓
[32/79] AUDIT_103 ✓
[33/79] AUDIT_104 ✓
[34/79] AUDIT_105 ✓
[35/79] AUDIT_106 ✓
[36/79] AUDIT_107 ✓
[37/79] AUDIT_108 ✓
[38/79] AUDIT_109 ✓
[39/79] AUDIT_110 ✓
[40/79] AUDIT_111 ✓
[41/79] AUDIT_112 ✓
[42/79] AUDIT_113 ✓
[43/79] AUDIT_114 ✓
[44/79] AUDIT_115 ✓

STOPPED at AUDIT_116
Error code: 429 - {'error': {'message': 'Rate li

In [58]:
resume_results_df = run_remaining_judges(
    pending_df=pending_df,
    output_path=resume_output_path,
    delay_seconds=0.2
)

Existing resume results: 87
Remaining API calls needed: 35
[1/35] AUDIT_116 ✓
[2/35] AUDIT_117 ✓
[3/35] AUDIT_118 ✓
[4/35] AUDIT_119 ✓
[5/35] AUDIT_120 ✓
[6/35] AUDIT_121 ✓
[7/35] AUDIT_122 ✓
[8/35] AUDIT_123 ✓
[9/35] AUDIT_124 ✓
[10/35] AUDIT_125 ✓
[11/35] AUDIT_126 ✓
[12/35] AUDIT_127 ✓
[13/35] AUDIT_128 ✓
[14/35] AUDIT_129 ✓
[15/35] AUDIT_130 ✓
[16/35] AUDIT_131 ✓
[17/35] AUDIT_132 ✓
[18/35] AUDIT_133 ✓
[19/35] AUDIT_134 ✓
[20/35] AUDIT_135 ✓
[21/35] AUDIT_136 ✓
[22/35] AUDIT_137 ✓
[23/35] AUDIT_138 ✓
[24/35] AUDIT_139 ✓
[25/35] AUDIT_140 ✓
[26/35] AUDIT_141 ✓
[27/35] AUDIT_142 ✓
[28/35] AUDIT_143 ✓
[29/35] AUDIT_144 ✓
[30/35] AUDIT_145 ✓
[31/35] AUDIT_146 ✓
[32/35] AUDIT_147 ✓
[33/35] AUDIT_148 ✓
[34/35] AUDIT_149 ✓
[35/35] AUDIT_150 ✓

RESUME RUN STATUS
Saved successful judgments: 122
Output: d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\llm_judge_resume_results.csv


In [59]:
import pandas as pd
from pathlib import Path
import re
import json

checkpoint_path = Path(
    r"d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\llm_judge_checkpoint_28.csv"
)

resume_path = Path(
    r"d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\llm_judge_resume_results.csv"
)

checkpoint = pd.read_csv(checkpoint_path)
resume = pd.read_csv(resume_path)

# Combine both judge result sources
judge_complete = pd.concat(
    [checkpoint, resume],
    ignore_index=True
)

# Normalize IDs
judge_complete["benchmark_id"] = (
    judge_complete["benchmark_id"].astype(str).str.strip()
)

expected_ids = set(
    benchmark_df["benchmark_id"].astype(str).str.strip()
)

actual_ids = set(judge_complete["benchmark_id"])

duplicates = (
    judge_complete["benchmark_id"]
    [judge_complete["benchmark_id"].duplicated()]
    .unique()
    .tolist()
)

missing_ids = sorted(expected_ids - actual_ids)
extra_ids = sorted(actual_ids - expected_ids)

# Parse validation
parsed_rows = []

for _, row in judge_complete.iterrows():
    parsed = parse_judge_json(row["judge_output"])
    parsed["benchmark_id"] = row["benchmark_id"]
    parsed_rows.append(parsed)

parsed_df = pd.DataFrame(parsed_rows)

score_columns = [
    "correctness",
    "relevance",
    "grounding",
    "helpfulness",
    "tone",
    "completeness"
]

invalid_scores = {}

for col in score_columns:
    invalid_scores[col] = int(
        (~parsed_df[col].isin([0, 1, 2])).sum()
    )

parse_failures = int(
    (parsed_df["parse_status"] != "SUCCESS").sum()
)

empty_outputs = int(
    judge_complete["judge_output"].isna().sum()
    + (judge_complete["judge_output"].astype(str).str.strip() == "").sum()
)

print("=" * 60)
print("FINAL LLM JUDGE INTEGRITY CHECK")
print("=" * 60)

print(f"Checkpoint rows:       {len(checkpoint)}")
print(f"Resume rows:           {len(resume)}")
print(f"Combined rows:         {len(judge_complete)}")
print(f"Unique benchmark IDs:  {judge_complete['benchmark_id'].nunique()}")
print(f"Expected IDs:          {len(expected_ids)}")

print("\nDuplicate IDs:", duplicates)
print("Missing IDs:", missing_ids)
print("Extra IDs:", extra_ids)

print("\nParse failures:", parse_failures)
print("Empty outputs:", empty_outputs)

print("\nInvalid score counts:")
for col, count in invalid_scores.items():
    print(f"  {col}: {count}")

print("\nID coverage complete:",
      actual_ids == expected_ids)

print("All rows parsed:",
      parse_failures == 0)

print("All scores valid:",
      all(v == 0 for v in invalid_scores.values()))

print("=" * 60)

FINAL LLM JUDGE INTEGRITY CHECK
Checkpoint rows:       28
Resume rows:           122
Combined rows:         150
Unique benchmark IDs:  150
Expected IDs:          150

Duplicate IDs: []
Missing IDs: []
Extra IDs: []

Parse failures: 0
Empty outputs: 0

Invalid score counts:
  correctness: 0
  relevance: 0
  grounding: 0
  helpfulness: 0
  tone: 0
  completeness: 0

ID coverage complete: True
All rows parsed: True
All scores valid: True


- combine the 150 judge results,
- parse all six dimensions,
- calculate mean score and % of maximum,
- calculate score distributions,
- calculate failure-mode frequencies,
- break results down by intent,
- break results down by escalation decision,
- create a final evaluation CSV for the report.

In [61]:
import pandas as pd
import numpy as np
import ast
from pathlib import Path

# ---------------------------------------------------------
# 1. Load the complete 150 judge results
# ---------------------------------------------------------

checkpoint = pd.read_csv(
    r"d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\llm_judge_checkpoint_28.csv"
)

resume = pd.read_csv(
    r"d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed\llm_judge_resume_results.csv"
)

judge_complete = pd.concat(
    [checkpoint, resume],
    ignore_index=True
)

judge_complete["benchmark_id"] = (
    judge_complete["benchmark_id"].astype(str).str.strip()
)

print("Judge results loaded:", len(judge_complete))
print("Unique IDs:", judge_complete["benchmark_id"].nunique())

# ---------------------------------------------------------
# 2. Parse judge outputs
# ---------------------------------------------------------

parsed_rows = []

for _, row in judge_complete.iterrows():
    parsed = parse_judge_json(row["judge_output"])
    parsed["benchmark_id"] = row["benchmark_id"]
    parsed_rows.append(parsed)

parsed_df = pd.DataFrame(parsed_rows)

# ---------------------------------------------------------
# 3. Merge with original benchmark
# ---------------------------------------------------------

evaluation_final = benchmark_df.merge(
    judge_complete[
        [
            "benchmark_id",
            "judge_output",
            "finish_reason",
            "judge_status"
        ]
    ],
    on="benchmark_id",
    how="inner",
    validate="one_to_one"
).merge(
    parsed_df,
    on="benchmark_id",
    how="inner",
    validate="one_to_one"
)

# ---------------------------------------------------------
# 4. Score columns
# ---------------------------------------------------------

score_columns = [
    "correctness",
    "relevance",
    "grounding",
    "helpfulness",
    "tone",
    "completeness"
]

for col in score_columns:
    evaluation_final[col] = pd.to_numeric(
        evaluation_final[col],
        errors="coerce"
    )

evaluation_final["overall_score"] = (
    evaluation_final[score_columns].mean(axis=1)
)

evaluation_final["overall_percentage"] = (
    evaluation_final["overall_score"] / 2.0 * 100
)

# ---------------------------------------------------------
# 5. Overall metrics
# ---------------------------------------------------------

overall_mean = evaluation_final[score_columns].mean()
overall_percentage = (overall_mean / 2.0) * 100

# ---------------------------------------------------------
# 6. Robust failure-flag normalization
# ---------------------------------------------------------

def normalize_failure_flags(value):

    # None / NaN
    if value is None:
        return []

    if isinstance(value, float) and np.isnan(value):
        return []

    # Already a list
    if isinstance(value, (list, tuple, set, np.ndarray)):
        result = []

        for item in value:
            if item is None:
                continue

            item = str(item).strip()

            if item and item != "NONE":
                result.append(item)

        return result

    # String representation of a list
    if isinstance(value, str):

        value = value.strip()

        if not value or value == "NONE":
            return []

        try:
            parsed = ast.literal_eval(value)

            if isinstance(parsed, (list, tuple, set)):
                return normalize_failure_flags(parsed)

        except Exception:
            pass

        return [value]

    return [str(value).strip()]


evaluation_final["failure_flags_normalized"] = (
    evaluation_final["failure_flags"]
    .apply(normalize_failure_flags)
)

# ---------------------------------------------------------
# 7. Failure-mode counts
# ---------------------------------------------------------

failure_counts = {}

for flags in evaluation_final["failure_flags_normalized"]:

    for flag in flags:

        if flag == "NONE" or not flag:
            continue

        failure_counts[flag] = (
            failure_counts.get(flag, 0) + 1
        )

failure_table = pd.DataFrame(
    [
        {
            "failure_mode": flag,
            "count": count,
            "percentage_of_150": count / len(evaluation_final) * 100
        }
        for flag, count in failure_counts.items()
    ]
)

if not failure_table.empty:
    failure_table = failure_table.sort_values(
        ["count", "failure_mode"],
        ascending=[False, True]
    ).reset_index(drop=True)

# ---------------------------------------------------------
# 8. Per-intent evaluation
# ---------------------------------------------------------

per_intent = (
    evaluation_final
    .groupby("intent_label")
    .agg(
        examples=("benchmark_id", "count"),
        overall_score=("overall_score", "mean"),
        correctness=("correctness", "mean"),
        relevance=("relevance", "mean"),
        grounding=("grounding", "mean"),
        helpfulness=("helpfulness", "mean"),
        tone=("tone", "mean"),
        completeness=("completeness", "mean")
    )
    .reset_index()
)

per_intent["overall_percentage"] = (
    per_intent["overall_score"] / 2.0 * 100
)

# ---------------------------------------------------------
# 9. By escalation policy
# ---------------------------------------------------------

by_policy = (
    evaluation_final
    .groupby("policy_decision")
    .agg(
        examples=("benchmark_id", "count"),
        overall_score=("overall_score", "mean"),
        correctness=("correctness", "mean"),
        relevance=("relevance", "mean"),
        grounding=("grounding", "mean"),
        helpfulness=("helpfulness", "mean"),
        tone=("tone", "mean"),
        completeness=("completeness", "mean")
    )
    .reset_index()
)

by_policy["overall_percentage"] = (
    by_policy["overall_score"] / 2.0 * 100
)

# ---------------------------------------------------------
# 10. Save final artifacts
# ---------------------------------------------------------

output_dir = Path(
    r"d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed"
)

evaluation_final.to_csv(
    output_dir / "llm_judge_evaluation_final_150.csv",
    index=False,
    encoding="utf-8-sig"
)

failure_table.to_csv(
    output_dir / "llm_judge_failure_modes_final.csv",
    index=False,
    encoding="utf-8-sig"
)

per_intent.to_csv(
    output_dir / "llm_judge_per_intent_final.csv",
    index=False,
    encoding="utf-8-sig"
)

by_policy.to_csv(
    output_dir / "llm_judge_by_policy_final.csv",
    index=False,
    encoding="utf-8-sig"
)

# ---------------------------------------------------------
# 11. Final output
# ---------------------------------------------------------

print("\n" + "=" * 70)
print("FINAL LLM-AS-JUDGE RESULTS — 150 EXAMPLES")
print("=" * 70)

print(f"\nBenchmark examples: {len(evaluation_final)}")

print("\nOverall dimension scores:")

for col in score_columns:
    print(
        f"{col:15s}: "
        f"{overall_mean[col]:.3f}/2 "
        f"({overall_percentage[col]:.1f}%)"
    )

print(
    f"\nOverall mean: "
    f"{evaluation_final['overall_score'].mean():.3f}/2 "
    f"({evaluation_final['overall_percentage'].mean():.1f}%)"
)

print("\nFailure modes:")

if failure_table.empty:
    print("No failure flags recorded.")
else:
    print(
        failure_table.to_string(index=False)
    )

print("\nPer-policy results:")
print(
    by_policy.round(3).to_string(index=False)
)

print("\nFiles saved:")
print(output_dir / "llm_judge_evaluation_final_150.csv")
print(output_dir / "llm_judge_failure_modes_final.csv")
print(output_dir / "llm_judge_per_intent_final.csv")
print(output_dir / "llm_judge_by_policy_final.csv")

print("=" * 70)

Judge results loaded: 150
Unique IDs: 150

FINAL LLM-AS-JUDGE RESULTS — 150 EXAMPLES

Benchmark examples: 150

Overall dimension scores:
correctness    : 1.707/2 (85.3%)
relevance      : 1.740/2 (87.0%)
grounding      : 1.693/2 (84.7%)
helpfulness    : 1.620/2 (81.0%)
tone           : 1.907/2 (95.3%)
completeness   : 1.680/2 (84.0%)

Overall mean: 1.724/2 (86.2%)

Failure modes:
          failure_mode  count  percentage_of_150
   INCOMPLETE_RESPONSE     20          13.333333
     FABRICATED_DETAIL     14           9.333333
     MISSED_ESCALATION     13           8.666667
    UNSUPPORTED_ACTION     13           8.666667
UNNECESSARY_ESCALATION      8           5.333333
   IRRELEVANT_RESPONSE      6           4.000000
          WRONG_INTENT      5           3.333333

Per-policy results:
policy_decision  examples  overall_score  correctness  relevance  grounding  helpfulness  tone  completeness  overall_percentage
    AUTO_HANDLE        90          1.819        1.789      1.856      1.756 

LLM-as-judge: On 150 stratified benchmark examples, the generated responses received a mean score of 1.724/2 (86.2%) across correctness, relevance, grounding, helpfulness, tone, and completeness. Tone was the highest-scoring dimension (95.3%), while helpfulness was lowest (81.0%). The most frequent flagged failures were incomplete responses (13.3%), fabricated details (9.3%), missed escalations (8.7%), and unsupported actions (8.7%). Because the same model family was used for response generation and judging, these results are treated as an automated evaluation signal rather than independent proof of quality.

In [64]:
from pathlib import Path
import pandas as pd

base = Path(r"d:\BECAME_DEVELOPER\hiver-sde-ai-agent\data\processed")

# ---------------------------------------------------------
# Load the 150-example benchmark
# ---------------------------------------------------------

benchmark = pd.read_csv(
    base / "response_generation_prompts_150.csv",
    encoding="utf-8-sig"
)

# ---------------------------------------------------------
# Load the actual Groq generation results
# ---------------------------------------------------------

groq = pd.read_csv(
    base / "groq_response_generation_150.csv",
    encoding="utf-8-sig"
)

print("Benchmark rows:", len(benchmark))
print("Groq generation rows:", len(groq))

print("\nGroq columns:")
print(groq.columns.tolist())

# ---------------------------------------------------------
# Select exactly 2 examples per intent
# ---------------------------------------------------------

selected = (
    benchmark
    .groupby("intent_label", group_keys=False)
    .sample(n=2, random_state=42)
)

human_eval = selected[
    [
        "benchmark_id",
        "intent_label",
        "policy_decision",
        "policy_reason_text",
        "customer_query",
        "generation_context",
        "historical_conversation"
    ]
].copy()

# ---------------------------------------------------------
# Identify the actual Groq response column
# ---------------------------------------------------------

response_candidates = [
    "groq_reply",
    "reply",
    "response",
    "generated_response",
    "assistant_response"
]

response_col = next(
    (c for c in response_candidates if c in groq.columns),
    None
)

if response_col is None:
    raise ValueError(
        "No response column found in Groq file.\n"
        f"Available columns:\n{groq.columns.tolist()}"
    )

print("\nDetected Groq response column:", response_col)

# ---------------------------------------------------------
# Merge actual generated response
# ---------------------------------------------------------

groq_response = groq[
    ["benchmark_id", response_col]
].copy()

groq_response = groq_response.rename(
    columns={response_col: "generated_response"}
)

human_eval = human_eval.merge(
    groq_response,
    on="benchmark_id",
    how="left",
    validate="one_to_one"
)

# ---------------------------------------------------------
# Validate
# ---------------------------------------------------------

if len(human_eval) != 30:
    raise ValueError(
        f"Expected 30 rows, got {len(human_eval)}"
    )

if human_eval["benchmark_id"].nunique() != 30:
    raise ValueError(
        "Duplicate benchmark IDs detected."
    )

if human_eval["generated_response"].isna().any():
    missing = human_eval.loc[
        human_eval["generated_response"].isna(),
        "benchmark_id"
    ].tolist()

    raise ValueError(
        f"Missing generated responses for: {missing}"
    )

# ---------------------------------------------------------
# Add human-rating columns
# ---------------------------------------------------------

for col in [
    "human_correctness",
    "human_relevance",
    "human_grounding",
    "human_helpfulness",
    "human_tone",
    "human_completeness",
    "human_failure_flags",
    "human_notes"
]:
    human_eval[col] = ""

# ---------------------------------------------------------
# Save
# ---------------------------------------------------------

output_path = base / "human_agreement_30.csv"

human_eval.to_csv(
    output_path,
    index=False,
    encoding="utf-8-sig"
)

print("\n" + "=" * 65)
print("HUMAN AGREEMENT SET READY")
print("=" * 65)

print("Rows:", len(human_eval))
print("Unique intents:", human_eval["intent_label"].nunique())
print("Generated responses:", human_eval["generated_response"].notna().sum())

print("\nIntent distribution:")
print(
    human_eval["intent_label"]
    .value_counts()
    .sort_index()
    .to_string()
)

print("\nSaved to:")
print(output_path)

print("=" * 65)

Benchmark rows: 150
Groq generation rows: 150

Groq columns:
['benchmark_id', 'conversation_id', 'intent_label', 'policy_decision', 'policy_reason_text', 'policy_rule', 'gemini_reply', 'groq_reply', 'provider', 'model', 'temperature', 'reasoning_effort', 'max_completion_tokens', 'finish_reason', 'reasoning_tokens', 'generation_status']

Detected Groq response column: groq_reply

HUMAN AGREEMENT SET READY
Rows: 30
Unique intents: 15
Generated responses: 30

Intent distribution:
intent_label
ACCOUNT_ACCESS_SECURITY             2
DELIVERY_ATTEMPT_OR_INSTRUCTIONS    2
DELIVERY_DELAY                      2
DELIVERY_MISSING_OR_MISDELIVERED    2
DEVICE_TECHNICAL_SUPPORT            2
DIGITAL_CONTENT                     2
GIFT_CARD_PROMOTION                 2
ORDER_STATUS_OR_CANCELLATION        2
OTHER_NON_SUPPORT                   2
PAYMENT_BILLING                     2
PRIME_MEMBERSHIP                    2
PRODUCT_AVAILABILITY_INFORMATION    2
PRODUCT_PROBLEM                     2
RETURN_REPL

Independent second-pass evaluation: A 30-example stratified subset was independently reviewed using the same 0–2 rubric used by the LLM judge. This second-pass evaluation was used as a consistency check and failure-analysis aid; it is not presented as human agreement because the ratings were produced by an independent evaluator rather than a separate human annotator

| Component | Evaluation | Result |
|---|---|---:|
| Intent: Majority baseline | 75-example dev split | 22.67% accuracy / 2.46% macro-F1 |
| Intent: TF-IDF + Logistic Regression | 75-example dev split | 17.33% accuracy / 15.29% macro-F1 |
| Intent: Customer-only TF-IDF | 75-example dev split | 25.33% accuracy / 12.64% macro-F1 |
| Intent: Word+char TF-IDF + SVM | 75-example dev split | 26.67% accuracy / 9.91% macro-F1 |
| Intent: Sentence embeddings + LR | 75-example dev split | 20.00% accuracy / 18.65% macro-F1 |
| Retrieval | 150 human-judged queries | 64.0% P@1 / 82.0% Hit@3 / 0.714 MRR |
| Generation | 150 examples, Groq | 1.724 / 2 overall LLM-judge score |
| Generation correctness | 150 examples | 1.707 / 2 |
| Generation relevance | 150 examples | 1.740 / 2 |
| Generation grounding | 150 examples | 1.693 / 2 |
| Generation helpfulness | 150 examples | 1.620 / 2 |
| Generation tone | 150 examples | 1.907 / 2 |
| Generation completeness | 150 examples | 1.680 / 2 |
| Escalation policy | 450 benchmark examples | 72.49% F1 |

The headline generation score of 1.724/2 (86.2%) should not be interpreted as an 86.2% real-world support-resolution rate. The evaluation uses a 150-example benchmark that was intentionally stratified by conversation size rather than sampled to represent the full AmazonHelp distribution. The response quality is also evaluated by an LLM judge rather than a production support outcome. In addition, the judge and generator use the same GPT-OSS-20B model family, which introduces potential self-evaluation bias. Finally, several responses received strong tone/relevance scores while still containing unsupported operational claims. Therefore, the headline score is best interpreted as a controlled benchmark of response quality—not as evidence that 86.2% of real customer issues would be safely resolved automatically.

The escalation evaluation measures agreement with a project-defined target policy rather than a human-verified escalation ground truth. The resulting F1 of 72.49% therefore measures consistency with the defined policy, not production escalation correctness.

Dataset → brand → conversations → taxonomy → baselines → classifier → retrieval → escalation → generation → LLM evaluation → independent review → failure analysis.